Add additional checks to optimization of constant static field loads

BruceForstall opened the pull request:
In `fgGetStaticFieldSeqAndAddress`, if we have a static field address altered by a tree of `ADD CNS_INS` nodes, we need to verify that the address is within the found field sequence. It might not be after shared constant CSE kicks in (e.g., under OptRepeat), where the sequence of ADDs might alter an arbitrary constant address from one type into the address of the static field of a different type. So we can't use the FieldSeq of the base address when considering the full offset.

BruceForstall committed:
Add additional checks to optimization of constant static field loads

In `fgGetStaticFieldSeqAndAddress`, if we have a static field address
altered by a tree of `ADD CNS_INS` nodes, we need to verify that the
address is within the found field sequence. It might not be after
shared constant CSE kicks in (e.g., under OptRepeat), where the
sequence of ADDs might be the alter an arbitrary constant address from
one type into the address of the static field of a different type.
So we can't use the FieldSeq of the base address when considering
the full offset.

BruceForstall commented:
Found as part of https://github.com/dotnet/runtime/pull/94250

BruceForstall commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -11061,11 +11064,26 @@ static bool GetStaticFieldSeqAndAddress(ValueNumStore* vnStore, GenTree* tree, s
-            *byteOffset = vnStore->CoercedConstantValue<ssize_t>(treeVN) - fldSeq->GetOffset() + val;
-            return true;
+
+            // Is the offset within the field? There might be a sequence of adds on top of a constant static field
+            // address due to shared constant CSE such that `treeVN` is not the static field base address.
+            // Check if the entire access doesn't span across the field and any other field. This simple check is
+            // just that the access starts within the field.
+            CORINFO_FIELD_HANDLE fieldHandle = fldSeq->GetFieldHandle();
+            CORINFO_CLASS_HANDLE fieldClsHnd;
+            var_types            fieldType = eeGetFieldType(fieldHandle, &fieldClsHnd);
+            ssize_t              fieldSize = (fieldType == TYP_STRUCT) ? info.compCompHnd->getClassSize(fieldClsHnd)
+                                                          : (ssize_t)genTypeSize(fieldType);
+            ssize_t tmpByteOffset = vnStore->CoercedConstantValue<ssize_t>(treeVN) - fldSeq->GetOffset() + val;
+            if ((0 <= tmpByteOffset) && (tmpByteOffset < fieldSize))
+            {
+                *pFseq      = fldSeq;
+                *byteOffset = tmpByteOffset;
+                return true;
+            }
```
This is the real fix. The rest is just clean-ups.

EgorBo replied:
I presume this the problem that is reproducable via Shared Constant + JitOptRepeat

I still don't think I understand the problem and it feels like this code is trying to mask some other problem or it's just some invalid IL that is doing something semi-legal, like reading out of bounds of a field (in that case `getStaticFieldContent` should just return false, if it currently hits asserts - we should probably just relax those, its contract is to return false on any non-foldable access)

The idea of this code was to simply return a field handle (as part of FieldSeq) + exact address of the content we need.

then, without validations, we ask VM via `getStaticFieldContent` to get us that content if it's valid and is immutable.

Typically, we work with trees like this:

```
IND(fieldAddr + offset) where fieldAddr has a field sequence (known static)
```
if the fieldAddr is split into pieces via shared constants, I presume the total offset should still be the same.

Maybe shared constants are doing something illegal around where exactly to leave a field sequence?

cc @jakobbotsch @SingleAccretion in case if this rings any bell

SingleAccretion replied:
Is a dump of where this failure occurs available?

The way the code is written (before this change), it effectively expects `getStaticFieldContent` to validate against nonsensical offsets that can occur in UB-like code.

If this occurs in not-UB-like code, that would be a bug somewhere else, not here.

BruceForstall replied:
There is no UB code here.

In one test case (Linux/arm32, 
```
System.Console:<get_IsInputRedirected>g__EnsureInitialized|34_0():System.Runtime.CompilerServices.StrongBox`1[ubyte] (MethodHash=e981a3d5)
```
with OptRepeat enabled (using all the fixes from https://github.com/dotnet/runtime/pull/94250), we get an assert:
```
Assert failure(PID 31115 [0x0000798b], Thread: 31115 [0x798b]): valueOffset == 0
    File: /home/brucefo/gh/runtime2/src/coreclr/vm/jitinterface.cpp:11783
```
https://github.com/dotnet/runtime/blob/8c68710084b861c44b6868c0bd972ce24aed55dd/src/coreclr/vm/jitinterface.cpp#L11780-L11791

We start with Value Numbering:
```
***** BB01, STMT00004(before)
N003 (  6, 11) [000012] ----G-------                        *  RETURN    ref
N002 (  5, 10) [000011] n---G-------                        \--*  IND       ref
N001 (  2,  8) [000010] H-----------                           \--*  CNS_INT(h) int    -0x4D1FE0A0 static Fseq[_isStdInRedirected]

N001 [000010]   CNS_INT(h) -0x4D1FE0A0 static Fseq[_isStdInRedirected] => $c4 {Hnd const: 0x0xb2e01f60 GTF_ICON_STATIC_HDL}
    VNForHandle(_isStdInRedirected) is $c5, fieldType is ref, size = 4
      AX1: select([$82]store($2c1, $c5, $101), $c5) ==> $101.
    VNForMapSelect($2c1, $c5):ref returns $101 {JitNew($c2, $141)}
N002 [000011]   IND       => <l:$101 {JitNew($c2, $141)}, c:$142 {MemOpaque:NotInLoop}>
N003 [000012]   RETURN    => $VN.Void
```

Because of shared constant CSE, we replace the `[000010]` tree by defining a CSE base address and use an offset, as follows:
```
N006 (  6, 11)              [000022] DA--G-------                        *  STORE_LCL_VAR ref    V02 tmp2         d:1 $VN.Void                                                                                                        N005 (  6, 11)              [000014] nA--G-------                        \--*  IND       ref    <l:$100, c:$140>                                                                                                                      N004 (  3,  9)              [000037] -A----------                           \--*  COMMA     int    $c0                                                                                                                                N002 (  2,  8) CSE #01 (def)[000035] DA----------                              +--*  STORE_LCL_VAR int    V04 cse0         d:1 $VN.Void                                                                                               N001 (  2,  8)              [000013] H-----------                              |  \--*  CNS_INT(h) int    -0x4D1FE0D0 static Fseq[STDIN_FILENO] $c0                                                                                   N003 (  1,  1)              [000036] ------------                              \--*  LCL_VAR   int    V04 cse0         u:1 $c0           
...
N005 (  5,  3)              [000012] ----G-------                        *  RETURN    ref    $VN.Void                                                                                                                                 N004 (  4,  2)              [000011] n---G-------                        \--*  IND       ref    <l:$101, c:$142>
N003 (  3,  4)              [000043] -------N----                           \--*  ADD       int    $c4
N001 (  1,  1)              [000041] ------------                              +--*  LCL_VAR   int    V04 cse0         u:1 $c0
N002 (  1,  2)              [000042] ------------                              \--*  CNS_INT   int    48           
```

In the 2nd iteration of JitOptRepeat, we re-run Value Numbering on:
```
***** BB01, STMT00004(before)
N005 (  5,  3) [000012] ----G-------                        *  RETURN    ref
N004 (  4,  2) [000011] n---G-------                        \--*  IND       ref
N003 (  3,  4) [000043] -------N----                           \--*  ADD       int
N001 (  1,  1) [000041] ------------                              +--*  LCL_VAR   int    V04 cse0         u:1 (last use)
N002 (  1,  2) [000042] ------------                              \--*  CNS_INT   int    48

N001 [000041]   LCL_VAR   V04 cse0         u:1 (last use) => $c0 {Hnd const: 0x0xb2e01f30 GTF_ICON_STATIC_HDL}
N002 [000042]   CNS_INT   48 => $46 {IntCns 48}
N003 [000043]   ADD       => $47 {IntCns 0xB2E01F60}                      
```
then hit the assert.

This is because `fgGetStaticFieldSeqAndAddress` walks past the `ADD` to find the FieldSeq for the `cse0` node, and sets the byteOffset to 48, from the `CNS_INT` node. This is, of course, non-sensical for the `cse0` field. `fgValueNumberConstLoad` then passes this on to `getStaticFieldContent` which asserts because the field type is REF and we're giving a non-zero offset.

The fix here adds some additional checking -- validation checking, if you will -- that the offset is valid and within the field.

An alternative might be to converts the asserts in `getStaticFieldContent` to `if` statements: https://github.com/dotnet/runtime/pull/100320

BruceForstall replied:
Here's the whole dump before the crash. [fail.txt](https://github.com/dotnet/runtime/files/14766077/fail.txt)

SingleAccretion replied:
Thank you - after looking at the dump I agree that this (or checking on the EE side) is the right fix. This is a case where the Jit itself creates an out-of-bounds field sequence access.

(Of course, for CQ purposes, it would be better if it didn't, but that's a different problem.)

BruceForstall commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -18928,9 +18926,10 @@ CORINFO_CLASS_HANDLE Compiler::gtGetFieldClassHandle(CORINFO_FIELD_HANDLE fieldH
-            JITDUMP("\nQuerying runtime about current class of field %s (declared as %s)\n",
-                    eeGetFieldName(fieldHnd, true, fieldNameBuffer, sizeof(fieldNameBuffer)),
-                    eeGetClassName(fieldClass, classNameBuffer, sizeof(classNameBuffer)));
+            eeGetFieldName(fieldHnd, true, fieldNameBuffer, sizeof(fieldNameBuffer));
+            eeGetClassName(fieldClass, classNameBuffer, sizeof(classNameBuffer));
```
The idea of pulling these calls out of the JITDUMP is that we'll get the actual field and class names in SPMI collections, so they will replay with JitDump to give us the actual names.

jakobbotsch replied:
I think you should put it under `EnableExtraSuperPmiQueries` to avoid the extra metadata queries (even under debug). E.g. `pickGDV` does the same:
https://github.com/dotnet/runtime/blob/c52c808300dee381369b6301b876ee177765ce31/src/coreclr/jit/importercalls.cpp#L6247-L6306

BruceForstall replied:
Good idea. Done.

BruceForstall commented:
/azp run runtime-coreclr outerloop, runtime-coreclr jitstress, runtime-coreclr libraries-jitstress

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -18928,9 +18926,10 @@ CORINFO_CLASS_HANDLE Compiler::gtGetFieldClassHandle(CORINFO_FIELD_HANDLE fieldH
-            JITDUMP("\nQuerying runtime about current class of field %s (declared as %s)\n",
-                    eeGetFieldName(fieldHnd, true, fieldNameBuffer, sizeof(fieldNameBuffer)),
-                    eeGetClassName(fieldClass, classNameBuffer, sizeof(classNameBuffer)));
+            eeGetFieldName(fieldHnd, true, fieldNameBuffer, sizeof(fieldNameBuffer));
+            eeGetClassName(fieldClass, classNameBuffer, sizeof(classNameBuffer));
```
The return pointer should be used (the input buffers are just used for scratch space).

BruceForstall committed:
Review feedback

1. Use `eeGetFieldName` / `eeGetClassName` return pointer
2. Only query extra metadata under `verbose || JitConfig.EnableExtraSuperPmiQueries()`

BruceForstall commented:
/azp run runtime-coreclr outerloop, runtime-coreclr jitstress, runtime-coreclr libraries-jitstress

BruceForstall commented:
/azp run runtime-coreclr outerloop, runtime-coreclr jitstress

BruceForstall commented:
Despite the GitHub "Checks" display above, all the tests completed and were successful, including [jitstress](https://dev.azure.com/dnceng-public/public/_build/results?buildId=605947&view=results).

[No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=605951&view=ms.vss-build-web.run-extensions-tab), although there are new SPMI "missing" contexts.

BruceForstall commented:
@egorbo PTAL
cc @dotnet/jit-contrib

EgorBo commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -18411,24 +18411,23 @@ bool Compiler::gtStoreDefinesField(
+    if (!tree->TypeIs(TYP_REF))
     {
-        return objClass;
+        return nullptr;
```
nit: we use NO_HANDLE instead of nullptr for CORINFO_CLASS_HANDLE

EgorBo reviewed (approved):
I am fine with merging this as is if it blocks JitOptRepeat, I just still think we should not break `GT_CNS_INT(h)` into pieces via SharedConstCSE and then do another iteration of JitOptRepeat untill we're sure it's perfectly fine to do
